src/Storages/StorageMerge.cpp

```diff
@@ -789,6 +846,83 @@ QueryPipelineBuilderPtr ReadFromMerge::createSources(
     return builder;
 }
 
+ReadFromMerge::RowPolicyData::RowPolicyData(RowPolicyFilterPtr row_policy_filter_ptr_,
+    std::shared_ptr<DB::IStorage> storage,
+    ContextPtr local_context)
+    : row_policy_filter_ptr(row_policy_filter_ptr_)
+{
+    assert(row_policy_filter_ptr_);
+
+    ASTPtr expr = row_policy_filter_ptr->expression;
+
+    auto storage_metadata_snapshot = storage->getInMemoryMetadataPtr();
+    auto storage_columns = storage_metadata_snapshot->getColumns();
+    auto needed_columns = storage_columns.getAllPhysical();
+
+    auto syntax_result = TreeRewriter(local_context).analyze(expr, needed_columns);
```

Comment:
Is it ok to use here `storage_columns.getAllPhysical()` as `needed_columns`? I mean what if a storage has an alias column and a row policy filter involves that alias?

Reply:
> Is it ok to use here `storage_columns.getAllPhysical()` as `needed_columns`? I mean what if a storage has an alias column and a row policy filter involves that alias? Can you please check that case?

Very good point, quite possible that it is a mistake. Will check.

Reply:
Indeed we have an issue here.
But the problem is deeper, Engine Merge is not very good at aliases in general, they seem do not work via function https://fiddle.clickhouse.com/4648fd3d-3c1a-462e-919b-90dafcafc131

Reply:
The problem is even deeper ;) , looks like Engine Merge does not work with tables with different structures. And 'does not work' means gives wrong result.
https://fiddle.clickhouse.com/27a3d1d1-51a4-491c-b6ad-5421c11d1f0b

Reply:
Summing up issue with alias.
I've fixed it for cases where Engine Merge is used directly, not via table function merge(). Works ok. Not pushed yet.
I haven't fixed it for merge() yet.
How it looks https://pastila.nl/?00d38392/d68be2ab03fb64572f5a3f1b3bd383b2#2RTThmq3iM6cf1C1nLQVUg==

Alternatively we can 

1. expose alias via merge table function - trivial fix. Can be treated as improvement, but breaks current behavior.
2. go with this PR further as is and fix alias based row policy for merge() in a separated PR.

Reply:
Considering we have  _asterisk_include_alias_columns_ the first option seems fine.
Currently table function merge() does not expose aliases.
@vitlibar , what do you think?

Reply:
> expose alias via merge table function - trivial fix. Can be treated as improvement, but breaks current behavior.

I suppose ideally the `merge()` function should return aliases only if that alias is requested explicitly; or if asterisk is used and ` asterisk_include_alias_columns` is set. Because it's better to be consistent with how we treat aliases for normal tables.

Reply:
Of course a better solution is always better, I just suggested to split work between two PRs. But ok, no problem, do it the way you like.

Reply:
The last push brought many changes in term of lines of code, but actually changes are not significant.
I've introduced processAliases method because I have to add new names before taking care of aliases and it is too much clutter for initializePipeline().
Besides this, I renamed convertingSourceStream to convertAndFilterSourceStream , because filtering is needed in-between.

Reply:
asterisk_include_alias_columns is left out of the scope despite my initial intention (and claim).